src/imputation/tmi_imputation.py

```diff
@@ -0,0 +1,337 @@
+import pandas as pd
+import math
+import numpy as np
+from src.staging.pg_conversion import sic_to_pg_mapper
+
+def apply_to_original(filtered_df, original_df):
+    original_df.loc[filtered_df.index] = filtered_df
+    return original_df
+
+
+def filter_by_column_content(
+    raw_df: pd.DataFrame, column: str, column_content: list
+) -> pd.DataFrame:
+
+    clean_df = raw_df[raw_df[column].isin(column_content)].copy()
+
+    return clean_df
+
+
+def impute_pg_by_sic(df: pd.DataFrame, sic_mapper: pd.DataFrame):
+    """Impute missing product groups for companies that do no r&d
+    and status = "Form sent out" using the SIC number ('rusic').
+
+    Args:
+        df (pd.DataFrame): initial dataframe from staging
+        sic_mapper (pd.DataFrame): SIC to PG mapper
+
+    Returns:
+        pd.DataFrame: Returns the original dataframe with new PGs
+        overwriting the missing values
+    """
+    
+    df['200'] = df['200'].astype("category")
+    
+    long_df = filter_by_column_content(df, "formtype", ["0001"])
+
+    # Filter for q604 = No or status = "Form sent out"
+    filtered_data = long_df.loc[
+        (long_df["status"] == "Form sent out") | (long_df["604"] == "No")
+    ]
+
+    filtered_data = sic_to_pg_mapper(filtered_data, sic_mapper, target_col="201", formtype = "0001")
+
+    updated_df = apply_to_original(filtered_data, df)
+
+    return updated_df
+
+
+# # Impute Business R&D type TODO: STEP 2
+# filtered_data["200"] = np.random.choice(
+#     ["C", "D"], size=len(filtered_data), p=[0.7, 0.3]
+# )
+# cur_df = apply_to_original(filtered_data, cur_df)
+
+
+def create_imp_class_col(
+    clean_df: pd.DataFrame, col_first_half: str, col_second_half: str, class_name: str
+) -> pd.DataFrame:
+
```

Comment:
create new dataframe to avoid changing original at line 59:
result_df = clean_df.copy()
--then exchange clean_df with results_df then at line 71 do:
result_df.update(fil_df)